src/extract/global_greedy_dag.rs

```diff
@@ -0,0 +1,193 @@
+use rpds::{HashTrieMap, HashTrieSet};
+
+use super::*;
+
+type TermId = usize;
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+struct Term {
+    op: String,
+    children: Vec<TermId>,
+}
+
+type Reachable = HashTrieSet<ClassId>;
+
+struct TermInfo {
+    node: NodeId,
+    eclass: ClassId,
+    node_cost: NotNan<f64>,
+    total_cost: NotNan<f64>,
+    // store the set of reachable terms from this term
+    reachable: Reachable,
+    size: usize,
+}
+
+// A TermDag needs to store terms that share common
+// subterms using a hashmap.
+// However, it also critically needs to be able to answer
+// reachability queries in this dag `reachable`.
+// This prevents double-counting costs when
+// computing the cost of a term.
+#[derive(Default)]
+pub struct TermDag {
+    nodes: Vec<Term>,
+    info: Vec<TermInfo>,
+    hash_cons: HashMap<Term, TermId>,
+}
+
+impl TermDag {
+    // Makes a new term using a node and children terms
+    // Correctly computes total_cost with sharing
+    // If this term contains itself, returns None
+    // If this term costs more than target, returns None
+    pub fn make(
+        &mut self,
+        node_id: NodeId,
+        node: &Node,
+        children: Vec<TermId>,
+        target: NotNan<f64>,
+    ) -> Option<TermId> {
+        let term = Term {
+            op: node.op.clone(),
+            children: children.clone(),
+        };
+
+        if let Some(id) = self.hash_cons.get(&term) {
+            return Some(*id);
+        }
+
+        let node_cost = node.cost;
+
+        if children.is_empty() {
+            let next_id = self.nodes.len();
+            self.nodes.push(term.clone());
+            self.info.push(TermInfo {
+                node: node_id,
+                eclass: node.eclass.clone(),
+                node_cost,
+                total_cost: node_cost,
+                reachable: [node.eclass.clone()].into_iter().collect(),
+                size: 1,
+            });
+            self.hash_cons.insert(term, next_id);
+            Some(next_id)
```

Comment:
optional: 

I think when the if/else are this long it can help to unnest after the "easy case", namely:

```rust
if children.is_empty() {
...
  return Some(next_id);
}
// check if children contains this node
for child in &children {
...
```

Reply:
eh, I like nesting
`return` considered harmful

Reply:
Up to you! I suggested this because there's already a good amount of early-return in this function (78, 92, 99, for example).

I personally think it is good sometimes, and not good other times. For code like this that feels pretty imperative either way I think it can make the code clearer. Feel free to keep as is though

Reply:
Ah I see
re-examining my head it seems that early return from a big code block makes me sad